src/game/client/components/infomessages.cpp

```diff
@@ -0,0 +1,570 @@
+/* (c) Magnus Auvinen. See licence.txt in the root of the distribution for more information. */
+/* If you are missing that file, acquire a complete release at teeworlds.com.                */
+#include <engine/graphics.h>
+#include <engine/shared/config.h>
+#include <engine/shared/protocol.h>
+#include <engine/textrender.h>
+#include <game/generated/client_data.h>
+#include <game/generated/protocol.h>
+#include <game/localization.h>
+
+#include "infomessages.h"
+#include <game/client/animstate.h>
+#include <game/client/gameclient.h>
+#include <game/client/prediction/entities/character.h>
+#include <game/client/prediction/gameworld.h>
+
+void CInfoMessages::OnWindowResize()
+{
+	for(auto &InfoMsg : m_aInfoMsgs)
+	{
+		DeleteTextContainers(&InfoMsg);
+	}
+}
+
+void CInfoMessages::OnReset()
+{
+	m_InfoMsgCurrent = 0;
+	for(auto &InfoMsg : m_aInfoMsgs)
+	{
+		InfoMsg.m_Tick = -100000;
+		DeleteTextContainers(&InfoMsg);
+	}
+}
+
+void CInfoMessages::DeleteTextContainers(CInfoMsg *pInfoMsg)
+{
+	TextRender()->DeleteTextContainer(pInfoMsg->m_VictimTextContainerIndex);
+	TextRender()->DeleteTextContainer(pInfoMsg->m_KillerTextContainerIndex);
+	TextRender()->DeleteTextContainer(pInfoMsg->m_DiffTextContainerIndex);
+	TextRender()->DeleteTextContainer(pInfoMsg->m_TimeTextContainerIndex);
+}
+
+void CInfoMessages::OnInit()
+{
+	Graphics()->SetColor(1.f, 1.f, 1.f, 1.f);
+	m_SpriteQuadContainerIndex = Graphics()->CreateQuadContainer(false);
+
+	Graphics()->QuadsSetSubset(0, 0, 1, 1);
+	RenderTools()->QuadContainerAddSprite(m_SpriteQuadContainerIndex, 0.f, 0.f, 28.f, 56.f);
+	Graphics()->QuadsSetSubset(0, 0, 1, 1);
+	RenderTools()->QuadContainerAddSprite(m_SpriteQuadContainerIndex, 0.f, 0.f, 28.f, 56.f);
+
+	Graphics()->QuadsSetSubset(1, 0, 0, 1);
+	RenderTools()->QuadContainerAddSprite(m_SpriteQuadContainerIndex, 0.f, 0.f, 28.f, 56.f);
+	Graphics()->QuadsSetSubset(1, 0, 0, 1);
+	RenderTools()->QuadContainerAddSprite(m_SpriteQuadContainerIndex, 0.f, 0.f, 28.f, 56.f);
+
+	for(int i = 0; i < NUM_WEAPONS; ++i)
+	{
+		float ScaleX, ScaleY;
+		Graphics()->QuadsSetSubset(0, 0, 1, 1);
+		RenderTools()->GetSpriteScale(g_pData->m_Weapons.m_aId[i].m_pSpriteBody, ScaleX, ScaleY);
+		RenderTools()->QuadContainerAddSprite(m_SpriteQuadContainerIndex, 96.f * ScaleX, 96.f * ScaleY);
+	}
+	Graphics()->QuadContainerUpload(m_SpriteQuadContainerIndex);
+}
+
+void CInfoMessages::AddInfoMsg(int Type, CInfoMsg NewMsg)
+{
+	NewMsg.m_Type = Type;
+	NewMsg.m_Tick = Client()->GameTick(g_Config.m_ClDummy);
+
+	m_InfoMsgCurrent = (m_InfoMsgCurrent + 1) % MAX_INFOMSGS;
+	DeleteTextContainers(&m_aInfoMsgs[m_InfoMsgCurrent]);
+	m_aInfoMsgs[m_InfoMsgCurrent] = NewMsg;
+}
+
+void CInfoMessages::CreateKillmessageNamesIfNotCreated(CInfoMsg *pInfoMsg)
```

Comment:
```suggestion
void CInfoMessages::CreateNamesIfNotCreated(CInfoMsg *pInfoMsg)
```
This seems to be called for both finish and kill messages.